hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala

```diff
@@ -103,13 +103,56 @@ final case class PNDArray(elementType: PType, nDims: Int, override val required:
     val outOfBounds = mb.newField[Boolean]
     Code(
       outOfBounds := false,
-      Code.foreach(0 until nDims){ dimIndex =>
+      Code.foreach(0 until nDims) { dimIndex =>
         outOfBounds := outOfBounds || (indices(dimIndex) >= shapeTuple(dimIndex))
       },
       outOfBounds
     )
   }
 
+  def linearizeIndices(indices: Array[Code[Long]], shapeArray: Array[Code[Long]], region: Code[Region], mb: MethodBuilder): Code[Long] = {
+    val index = mb.newField[Long]
+    val elementsInProcessedDimensions = mb.newField[Long]
+    Code(
+      index := 0L,
+      elementsInProcessedDimensions := 1L,
+      Code.foreach(shapeArray.zip(indices).reverse) { case (shapeElement, currentIndex) =>
+        Code(
+          index := index + currentIndex * elementsInProcessedDimensions,
+          elementsInProcessedDimensions := elementsInProcessedDimensions * shapeElement
+        )
+      },
+      index
+    )
+  }
+
+  def unlinearizeIndex(index: Code[Long], shapeArray: Array[Code[Long]], region: Code[Region], mb: MethodBuilder): (Code[Unit], Array[Code[Long]]) = {
+    val nDim = shapeArray.length
+
+    if (nDim <= 1) {
+      (Code._empty, Array(index))
```

Comment:
do we have a test for this case? it seems a bit weird that in every other case `unlinearizeIndex(...).length == shapeArray.length`

Reply:
I guess the question is, for a 0-dimensional ndarray, how many indices does `outputElement` expect? I'd think zero?